src/Services/Dominion/Actions/EspionageActionService.php

```diff
@@ -386,6 +409,204 @@ protected function performInfoGatheringOperation(Dominion $dominion, string $ope
         ];
     }
 
-    // todo: theft/black ops/war
+    protected function performResourceTheftOperation(Dominion $dominion, string $operationKey, Dominion $target): array
+    {
+        $operationInfo = $this->espionageHelper->getOperationInfo($operationKey);
+
+        $selfSpa = $this->militaryCalculator->getSpyRatio($dominion, 'offense');
+        $targetSpa = $this->militaryCalculator->getSpyRatio($target, 'defense');
+
+        // You need at least some positive SPA to perform espionage operations
+        if ($selfSpa === 0.0) {
+            return [
+                'success' => false,
+                'message' => "Your spy force is too weak to cast {$operationInfo['name']}. Please train some more spies.",
+                'alert-type' => 'warning',
+            ];
+        }
+
+        if ($targetSpa !== 0.0) {
+            $ratio = ($selfSpa / $targetSpa);
+
+            // todo: copied from spell success ratio. needs looking into later
+            // todo: factor in spy strength
+            $successRate = clamp((
+                (0.0172 * ($ratio ** 3))
+                - (0.1809 * ($ratio ** 2))
+                + (0.6767 * $ratio)
+                - 0.0134
+            ), 0.0, 1.0);
+
+            if (!random_chance($successRate)) {
+                // todo: move to CasualtiesCalculator
+
+                // Values (percentage)
+                $spiesKilledBasePercentage = 0.2; // TODO: Higher for black ops.
+                $forestHavenSpyCasualtyReduction = 3;
+                $forestHavenSpyCasualtyReductionMax = 30;
+
+                $spiesKilledMultiplier = (1 - min(
+                        (($dominion->building_forest_haven / $this->landCalculator->getTotalLand($dominion)) * $forestHavenSpyCasualtyReduction),
+                        ($forestHavenSpyCasualtyReductionMax / 100)
+                    ));
+
+                $spyLossSpaRatio = ($targetSpa / $selfSpa);
+                $spiesKilledPercentage = clamp($spiesKilledBasePercentage * $spyLossSpaRatio, 0.05, 0.5);
```

Comment:
```suggestion
                $spiesKilledPercentage = clamp($spiesKilledBasePercentage * $spyLossSpaRatio, 0.4, 2.5);
```

Formula for min/max caps on spy casualties% on failed spy ops:
- Max cap% = base% * 2.5
- Min cap%: base% / 2.5

For failed spy ops this would result in (0.08% min - 0.2% base - 0.5% max), although it's currently 0.05 - 0.2 - 0.5, so that needs changing.

Failed resource theft ops: 0.4% min - 1% base - 2.5% max.

For future reference: failed black ops: 0.8% min - 2% base = 5% max.

Reply:
Future note to self: Failed info ops is 0.1% casualties, not 0.2%